### PR TITLE
feat: disable unattended update for Mariner by default

### LINKED ISSUE #2257: Allow us to turn off Mariner's equivalent of unattended upgrade
**Is your feature request related to a problem?/Why is this needed**
We can turn off in ubuntu. If mariner has same functionality we should allow it to be turned off.
Don't think Windows has an equivalent. @AbelHu.

**Describe the solution you'd like in detail**
Honor NodeBootstrapingConfig's DisableUnattendedUpgrade and turn any automatic updates off. 

**Describe alternatives you've considered**
Daemonsets and remediators are fine but there is always a race with new images if we don't do it here.

**Additional context**

https://github.com/Azure/AgentBaker/pull/2175

### PR DESCRIPTION
**Pull request by dmcilvaney**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Unattended update has been disabled by default for Ubuntu, and re-enabled at provisioning time based on the cluster configuration.

This PR will bring Mariner into parity with Ubuntu.

**Which issue(s) this PR fixes**:

Fixes #2257

**Requirements**:
- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: parts/linux/cloud-init/artifacts/cse_main.sh [hunk 1]
```diff
@@ -336,6 +336,28 @@ else
             systemctl restart --no-block apt-daily.service
         fi
         aptmarkWALinuxAgent unhold &
+    elif [[ $OS == $MARINER_OS_NAME ]]; then
+        {{- if EnableUnattendedUpgrade }}
```
Comment (alexeldeib) [inline parts/linux/cloud-init/artifacts/cse_main.sh:340]: please avoid using go templates anywhere other than `cse_cmd.sh`, the goal is to remove them entirely.

can use `        if [ "${ENABLE_UNATTENDED_UPGRADES}" == "true" ]; then` here which is already plumbed for ubuntu if you rebase

File: parts/linux/cloud-init/artifacts/cse_main.sh [hunk 2]
```diff
@@ -336,6 +336,28 @@ else
...
+        {{- if IsKata }}
```
Comment (alexeldeib) [inline parts/linux/cloud-init/artifacts/cse_main.sh:341]: same, set this elsewhere

File: parts/linux/cloud-init/artifacts/cse_main.sh [hunk 3]
```diff
@@ -336,6 +336,28 @@ else
             systemctl restart --no-block apt-daily.service
         fi
         aptmarkWALinuxAgent unhold &
+    elif [[ $OS == $MARINER_OS_NAME ]]; then
```
Comment (alexeldeib) [inline parts/linux/cloud-init/artifacts/cse_main.sh:339]: if isMariner && !isKata removes the need for conditional below

Comment (dmcilvaney) [reply parts/linux/cloud-init/artifacts/cse_main.sh:339]: I was going to leave the echo that explains why we aren't turning on the updates in this specific case, but if we don't want that happy to tidy that up.

Comment (alexeldeib) [reply parts/linux/cloud-init/artifacts/cse_main.sh:339]: ah no the comments are fine. just the go templates was all

File: pkg/agent/testdata/MarinerV2+Kata/CSECommand [hunk 4]
```diff
@@ -1 +1 @@
...
...IiwKICAgICJzeXN0ZW1SZXNlcnZlZCI6IHsKICAgICAgICAiY3B1IjogIjIiLAogICAgICAgICJtZW1vcnkiOiAiMUdpIgogICAgfSwKICAgICJrdWJlUmVzZXJ2ZWQiOiB7CiAgICAgICAgImNwdSI6ICIxMDBtIiwKICAgICAgICAibWVtb3J5IjogIjE2MzhNaSIKICAgIH0sCiAgICAiZW5mb3JjZU5vZGVBbGxvY2F0YWJsZSI6IFsKICAgICAgICAicG9kcyIKICAgIF0KfQ==" SWAP_FILE_SIZE_MB="0" GPU_DRIVER_VERSION="cuda-510.47.03" GPU_INSTANCE_PROFILE="" CUSTOM_SEARCH_DOMAIN_NAME="" CUSTOM_SEARCH_REALM_USER="" CUSTOM_SEARCH_REALM_PASSWORD="" MESSAGE_OF_THE_DAY="" HAS_KUBELET_DISK_TYPE="false" NEEDS_CGROUPV2="false" SYSCTL_CONTENT="IyBUaGlzIGlzIGEgcGFydGlhbCB3b3JrYXJvdW5kIHRvIHRoaXMgdXBzdHJlYW0gS3ViZXJuZXRlcyBpc3N1ZToKIyBodHRwczovL2dpdGh1Yi5jb20va3ViZXJuZXRlcy9rdWJlcm5ldGVzL2lzc3Vlcy80MTkxNiNpc3N1ZWNvbW1lbnQtMzEyNDI4NzMxCm5ldC5pcHY0LnRjcF9yZXRyaWVzMj04Cm5ldC5jb3JlLm1lc3NhZ2VfYnVyc3Q9ODAKbmV0LmNvcmUubWVzc2FnZV9jb3N0PTQwCm5ldC5jb3JlLnNvbWF4Y29ubj0xNjM4NApuZXQuaXB2NC50Y3BfbWF4X3N5bl9iYWNrbG9nPTE2Mzg0Cm5ldC5pcHY0Lm5laWdoLmRlZmF1bHQuZ2NfdGhyZXNoMT00MDk2Cm5ldC5pcHY0Lm5laWdoLmRlZmF1bHQuZ2NfdGhyZXNoMj04MTkyCm5ldC5pcHY0Lm5laWdoLmRlZmF1bHQuZ2NfdGhyZXNoMz0xNjM4NAoK" TLS_BOOTSTRAP_TOKEN="" IS_KATA="true" /usr/bin/nohup /bin/bash -c "/bin/bash /opt/azure/containers/provision_start.sh"
```
Comment (alexeldeib) [inline pkg/agent/testdata/MarinerV2+Kata/CSECommand:1]: IS_KATA="true"


nice

File: vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh [hunk 5]
```diff
@@ -72,21 +72,29 @@ listInstalledPackages() {
...
-    systemctlEnableAndStart check-restart.timer || exit $ERR_SYSTEMCTL_START_FAIL
+    systemctl disable dnf-automatic-install.service || exit 1
+    systemctl mask dnf-automatic-install.service || exit 1
+
+    systemctl stop dnf-automatic-install.timer || exit 1
+    systemctl disable dnf-automatic-install.timer || exit 1
+    systemctl mask dnf-automatic-install.timer || exit 1
+}
+
+# Regardless of UU mode, ensure check-restart is running
```
Comment (alexeldeib) [inline vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh:90]: jfyi, would likely eventually like to get rid of tools files. some weird duplication that is very annoying. 

distro helpers will be baked into the vhd, and still accessible.

fine to leave in this pr

**Review by alexeldeib — APPROVED** [review]
lgtm, I may cause you a tiny bit more rebase pain

File: vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh [hunk 6]
```diff
@@ -72,21 +72,29 @@ listInstalledPackages() {
...
 }
 
-# By default the dnf-automatic is service is notify only in Mariner.
-# Enable the automatic install timer and the check-restart timer.
-enableDNFAutomatic() {
-    # Stop the notify only dnf timer since we've enabled the auto install one.
+# disable and mask all UU timers/services
+disableDNFAutomatic() {
+    # Make sure dnf-automatic is running with the notify timer rather than the auto install timer
+    systemctlEnableAndStart  dnf-automatic-notifyonly.timer || exit $ERR_SYSTEMCTL_START_FAIL
```
Comment (hbeberman) [inline vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh:78]: nit: double space

File: parts/linux/cloud-init/artifacts/cse_main.sh [hunk 7]
```diff
@@ -354,6 +354,26 @@ else
...
+                echo 'EnableUnattendedUpgrade is not supported by kata images, will not be enabled'
+            else
+                # By default the dnf-automatic is service is notify only in Mariner.
+                # Enable the automatic install timer and the check-restart timer.
+                # Stop the notify only dnf timer since we've enabled the auto install one.
+                # systemctlDisableAndStop adds .service to the end which doesn't work on timers.
+                systemctl disable dnf-automatic-notifyonly.timer
+                systemctl stop dnf-automatic-notifyonly.timer
+                # At 6:00:00 UTC (1 hour random fuzz) download and install package updates.
+                systemctl unmask dnf-automatic-install.service || exit $ERR_SYSTEMCTL_START_FAIL
```
Comment (hbeberman) [inline parts/linux/cloud-init/artifacts/cse_main.sh:371]: You `systemctl disable dnf-automatic-install.service` in the initial VHD setup, just confirming if you don't need to `systemctl enable dnf-automatic-install.service` here to undo that, or is the timer still able to start the service once its unmasked.

**Review by hbeberman — APPROVED** [review]

**Review by SeanDougherty — APPROVED** [review]

File: parts/linux/cloud-init/artifacts/cse_main.sh (hunk 7 above)
Comment (alexeldeib) [reply parts/linux/cloud-init/artifacts/cse_main.sh:371]: I think this works as long as you start the timer, I had a similar issue for ubuntu

Comment (alexeldeib) [reply parts/linux/cloud-init/artifacts/cse_main.sh:371]: pasting from a bit above:
```
      systemctl unmask apt-daily.service apt-daily-upgrade.service
      systemctl enable apt-daily.service apt-daily-upgrade.service
      systemctl enable apt-daily.timer apt-daily-upgrade.timer
```

**Comment by dmcilvaney** [comment]
Rebased to pull latest changes

File: parts/linux/cloud-init/artifacts/cse_main.sh (hunk 7 above)
Comment (dmcilvaney) [reply parts/linux/cloud-init/artifacts/cse_main.sh:371]: My understanding is that a timer triggered service is not actually started, but only triggered by the timer

**Review by alexeldeib — APPROVED** [review]
re-lgtm

**Review by hbeberman — APPROVED** [review]

**Pull request merged**